drivers/include/net/netdev2/ppp.h

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright (C) 2016 Fundación Inria Chile
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup drivers_netdev_netdev2
+ * @brief
+ * @{
+ *
+ * @file
+ * @brief   Definitions for netdev2 common Point to Point Protocol code
+ *
+ * @author  José Ignacio Alamos
+ */
+#ifndef NETDEV2_PPP_H
+#define NETDEV2_PPP_H
+
+#include "net/gnrc/ppp/prot.h"
+#include "net/gnrc/ppp/prot.h"
+#include "net/gnrc/ppp/lcp.h"
+#include "net/gnrc/ppp/pap.h"
+#include "net/gnrc/ppp/ipcp.h"
+#include "net/gnrc/nettype.h"
+#include "net/netopt.h"
+#include "net/netdev2.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#if defined(MODULE_GNRC_PPP) || doxygen
+/**
+ * @brief class of custom driver control protocol
+ * @extends ppp_protocol_t
+ *
+ * @details the DCP is in charge of monitoring the link and exchanging messages with the ppp device
+ */
+typedef struct gnrc_ppp_dcp {
+    gnrc_ppp_protocol_t prot;    /**< base ppp_protocol_t object */
+    msg_t timer_msg;        /**< msg struct for handling timeouts messages */
+    xtimer_t xtimer;        /**< xtimer struct for sending timeout messages */
+    uint8_t dead_counter;   /**< when reaches zero, the link is assumed to be dead */
+} gnrc_ppp_dcp_t;
+#endif
+
+
+/**
+ * @brief Extended structure to hold PPP driver state
+ *
+ * @extends netdev2_t
+ *
+ * Supposed to be extended by driver implementations.
+ * The extended structure should contain all variable driver state.
+ */
+typedef struct {
+    netdev2_t netdev;                       /**< @ref netdev2_t base class */
+#if defined(MODULE_GNRC_PPP) || doxygen
+    /**
+     * @brief PPP specific fields
+     * @{
+     */
+    gnrc_ppp_dcp_t dcp;                             /**< Control protocol for driver */
+    gnrc_ppp_lcp_t lcp;                             /**< Link Control Protocol */
+    gnrc_ppp_pap_t pap;                             /**< Password Authentication Protocol */
+    gnrc_ppp_ipcp_t ipcp;                  /**< IPv4 Network Control Protocol */
+    gnrc_ppp_ipv4_t ipv4;                  /**< Handler for IPv4 packets */
```

Comment:
You cast these to `gnrc_ppp_protocol_t` quite often in your code (https://github.com/RIOT-OS/RIOT/pull/5470/files#diff-6d2b26dab0a8a164f332c85c0910c139R35), but e.g. `sizeof(gnrc_ppp_dcp_t) < sizeof(gnrc_ppp_protocol)`, so you will access data after `netdev2_ppp_t.dcp` (something in `netdev2_ppp_t.lcp` in this case).